Add Blue-Green Deployment using K8s (#317)
tektoncd/catalog: blue-green-deploy/README.md

# Blue Green Deploy on Kubernetes

> This task gives you a flexibility to deploy the newer version of the application in the green zone without disturbing the previous version by making the service point to the newer version of the application deployed. It gives a rapid way to rollback - if anything goes wrong you change the service to point back to your previous deployment.

Reviewer: ```suggestion
This task gives you the flexibility to deploy the a version of the application from the blue zone to the green zone without disturbing the previous version in the blue zone, it works by making the service point to the newer version of the application deployed. It gives a rapid way to rollback - if anything goes wrong you change the service to point back to your previous deployment.
```

> THe following task can help you to deploy an application using the Blue-Green deployment strategy. This task gives you the flexibility to deploy a version of the application from the blue zone to the green zone without disturbing the previous version in the blue zone. It works by making the service point to the newer version of the application deployed. 

Reviewer 2: ```suggestion
The following task can help you to deploy an application using the Blue-Green deployment strategy. This task gives you the flexibility to deploy a version of the application from the blue zone to the green zone without disturbing the previous version in the blue zone. It works by making the service point to the newer version of the application deployed. 
```

> - If you are deploying the first version of the application then it will go into the blue zone and for that you need to provide deployment and service manifests via the `workspaces` and in the `params` the service name and version. Sample Kubernetes manifests for version v1 can be found [here](./v1-deploy).

Reviewer 2: Link is not working
I think it should be 
```suggestion
- If you are deploying the first version of the application then it will go into the blue zone and for that you need to provide deployment and service manifests via the `workspaces` and in the `params` the service name and version. Sample Kubernetes manifests for version v1 can be found [here](./examples/v1-deploy).
```

> - For further deployments the deployments will get deployed in the other zone to which the current service is pointing to. For example if we have a deployment running in the blue zone then we will deploy the next deployment in the green zone and make the service point to the green zone or vice versa.

Reviewer 2: `For further deployments the deployments will get deployed in the other zone to which the current service is pointing to.`
is that current service points to new deployment ?

## Installing the Task

```
kubectl apply -f https://raw.githubusercontent.com/tektoncd/catalog/master/blue-green-deploy/blue-green-deploy.yaml
```

## Installing the ClusterRoleBinding

```
kubectl apply -f https://raw.githubusercontent.com/tektoncd/catalog/master/blue-green-deploy/clusterrolebinding.yaml
```

## Workspaces

- **manifest-dir**: Manifest files can be provided via the workspaces.(Default: _emptyDir:{}_ in case no manifest is provided)
- **kubeconfig-dir**: If you want to deploy you application to another cluster then you can mount your `kubeconfig` file via this `workspace`. (Default: _emptyDir:{}_ in case `kubeconfig` is not mounted)

## Parameters

- **SERVICE_NAME**: The service name pointing to the existing deployment. (_Note_: The service name for the new deployment should be same)
- **NEW_VERSION**: The version of the deployment to be deployed in the green/blue zone

> - **MANIFEST**: The deployment manifest URL file path provided in case the manifest is present on Github. (_Example_: "https://raw.githubusercontent.com/vinamra28/blue-green-deployment-k8s/master/deployment%2Bservice/blue-deployment.yaml")

Reviewer 2: ```suggestion
- **MANIFEST**: The deployment manifest URL file path provided in case the manifest is present on Github. (_Example_: "https://raw.githubusercontent.com/tektoncd/catalog/master/blue-green-deploy/example/v1-deploy/blue-deployment.yaml")
```

# Usage

This TaskRun runs the Task to deploy the given Kubernetes resource in the green/blue zone and toggle the service to point to the new zone.

## Without using ConfigMap

TaskRun :-

```
apiVersion: tekton.dev/v1beta1
kind: TaskRun
metadata:
  name: blue-green-deploy-run
spec:
  taskRef:
    name: blue-green-deploy-k8s
  params:
    - name: SERVICE_NAME
      value: myapp
    - name: NEW_VERSION
      value: v2
    - name: MANIFEST

>       value: "https://raw.githubusercontent.com/vinamra28/blue-green-deployment-k8s/master/deployment%2Bservice/green-deployment.yaml"

Reviewer 2: ```suggestion
      value: "https://raw.githubusercontent.com/tektoncd/catalog/master/blue-green-deploy/example/v2-deploy/green-deployment.yamll"
```

# Blue Green Deploy on Kubernetes

This task gives you a flexibility to deploy the newer version of the application in the green zone without disturbing the previous version by making the service point to the newer version of the application deployed. It gives a rapid way to rollback - if anything goes wrong you switch the router back to your blue environment.

  - If you are deploying version `v1` of the application then you need to provide deployment and service manifests via the `ConfigMap` and in the `params` the service name and version. Sample Kubernetes  manifests for version v1 can be found [here](./v1-deploy).
  - If you are deploying version `v2` of the application the you need to provide only the deployment manifest and in the `params` the new deployment service name and version. Sample Kubernetes  manifests for version v2 can be found [here](./v2-deploy).

> ![](./blue_green_deployments.png)

Reviewer: What is this diagram refer too ? I see some green and blue blobs and arrow between the green blobs but i am not sure what is trying to explains!

# Blue Green Deploy on Kubernetes

## Introduction

One of the challenges with automating deployment is the cut-over itself, taking software from the final stage of testing to live production. You usually need to do this quickly in order to minimize downtime. The blue-green deployment approach does this by ensuring you have two production environments, as identical as possible. At any time one of them, let's say blue for the example, is live. As you prepare a new release of your software you do your final stage of testing in the green environment. Once the software is working in the green environment, you switch the router so that all incoming requests go to the green environment - the blue one is now idle.

> Blue-green deployment also gives you a rapid way to rollback - if anything goes wrong you switch the router back to your blue environment.

Reviewer: So we have the "marketting" description on how blue green deployment works but it would be better i think to explains how the task actually works? I would not know how it works unless i start to read the shell script....
